### PR TITLE
[CI] Make nightly builds with Github Actions CI

### PR DESCRIPTION
**Pull request by aelovikov-intel**
They way release action works is that it overrides the release if ran multiple times but doesn't update the tag. As such, manual workflow runs could present a challenge. To deal with that, I also use short commit hash as a suffix to ensure tag and release are in sync unless the triggering event is schedule (which only runs once a day and doesn't result in overwrite issue).

### CODE REVIEW

**Comment by aelovikov-intel** [comment]
Tested in:
https://github.com/intel/llvm/actions/runs/5660054834/job/15335100289 - original version (1st commit in this PR's branch), verify E2E tests work after modifications.
https://github.com/intel/llvm/actions/runs/5661054271/job/15338445579 - reduced E2E testing (to reduce time), verify the work of the release part.

File: .github/workflows/sycl_nightly.yml [hunk 1]
```diff
@@ -27,6 +27,12 @@ jobs:
       lts_matrix: ${{ needs.test_matrix.outputs.lts_lx_matrix }}
       lts_aws_matrix: ${{ needs.test_matrix.outputs.lts_aws_matrix }}
 
+      # We use the build for the release, override its name and prefer
+      # widespread gzip compression.
+      artifact_archive_name: sycl_linux.tar.gz
```
Comment (bader) [inline .github/workflows/sycl_nightly.yml:32]:
```suggestion
      artifact_archive_name: dpcpp.lin.tgz
```
Technically the name of the compiler is oneAPI DPC++. What does `sycl` refer to? Branch name?
Mercy for my fingers, please! `.` is easier to type than `_`.

I think release version must be added to the file name, so we can keep multiple releases in the same directory.

File: .github/workflows/sycl_nightly.yml [hunk 2]
```diff
@@ -48,6 +54,41 @@ jobs:
       retention-days: 90
       lts_matrix: ${{ needs.test_matrix.outputs.lts_wn_matrix }}
 
+      # We use the build for the release, override its name.
+      artifact_archive_name: sycl_windows.tar.gz
```
Comment (bader) [inline .github/workflows/sycl_nightly.yml:58]:
```suggestion
      artifact_archive_name: dpcpp.win.tgz
```
I think usually tar is not used on windows. Can we just use (self-extracting?) zip or Windows installer to make handling easier for Windows users?

**Review by stdale-intel — APPROVED** [review]
Looks good, all pieces in place. great work @aelovikov-intel !

File: .github/workflows/sycl_nightly.yml [hunk 3]
```diff
@@ -27,6 +27,12 @@ jobs:
       lts_matrix: ${{ needs.test_matrix.outputs.lts_lx_matrix }}
       lts_aws_matrix: ${{ needs.test_matrix.outputs.lts_aws_matrix }}
 
+      # We use the build for the release, override its name and prefer
+      # widespread gzip compression.
```
Comment (bader) [inline .github/workflows/sycl_nightly.yml:31]: For instance, LLVM uses LZMA - https://github.com/llvm/llvm-project/releases/tag/llvmorg-16.0.0.

Is `gzip` "widespread" on Windows too?

File: .github/workflows/sycl_nightly.yml [hunk 4]
```diff
@@ -48,6 +54,41 @@ jobs:
...
+        name: sycl_windows_default
+    - name: Compute tag
+      id: tag
+      run: |
+        if [ "${{ github.event_name == 'schedule' }}" == "true" ]; then
+          echo "TAG=$(date +'%Y-%m-%d')" >> "$GITHUB_OUTPUT"
+        else
+          # TODO: Use date of the commit?
+          echo "TAG=$(date +'%Y-%m-%d')-${GITHUB_SHA::7}" >> "$GITHUB_OUTPUT"
+        fi
```
Comment (bader) [inline .github/workflows/sycl_nightly.yml:80]: What is the purpose of a "nightly" release?
For regular releases, we use semantic versioning, which should work pretty well for nightly releases too.
What is the problem with using SHA for scheduled releases as well?

**Comment by aelovikov-intel** [comment]
@bader , regarding your questions about "tar", "gzip" and "nightly":

We already have nightly releases under https://github.com/intel/llvm/releases/tag/sycl-nightly%2F20230725 but that's being done outside Github CI. This PR transitions those tasks into in-tree CI. It seems current consumers of those releases are satisfied with current packing/naming, so I'm not sure if it's worth changing it.

> we use semantic versioning

I don't think that would apply to those builds (as opposite to official release schedule). We only bump major version once through the release cycle, even if there are multiple ABI/API breaking PRs during that development window.

> What is the problem with using SHA for scheduled releases as well?

There aren't any technical problems, but not having the short SHA might look nicer.

That said, I'm open to changing naming/archiving/tagging scheme, but we'd need input from actual customers who use those builds. @stdale-intel , do you know who we should tag here?

**Comment by stdale-intel** [comment]
> I don't think that would apply to those builds (as opposite to official release schedule). We only bump major version once through the release cycle, even if there are multiple ABI/API breaking PRs during that development window.

I agree with this and don't think we should have semantic versioning on nightly ci builds. Nightly builds should be just that, a build of that days changes and mostly likely just be versioned with the build date for easy reference (sycl.linux.072523) as they as solely intended long term to be quick and easy way to test that days changes (and not something to align to a specific version/release). 

>That said, I'm open to changing naming/archiving/tagging scheme, but we'd need input from actual customers who use those builds. @stdale-intel , do you know who we should tag here?

For now, i'd prefer us not make any drastic changes as part of this shift to building nightly builds with github actions vs how we were doing it before. We have other upcoming changes we have planned there will be chance to review all of this in a broader sense with customer feedback (FYI @aelovikov-intel  we want blender/render folks, mkl, dnn, dl frameworks folks input in terms of customers) that I think fits better than trying to resolve it for this PR.

**Review by offsake — COMMENTED** [review]
LGTM.

File: .github/workflows/sycl_nightly.yml [hunk 5]
```diff
@@ -48,6 +52,42 @@ jobs:
       retention-days: 90
       lts_matrix: ${{ needs.test_matrix.outputs.lts_wn_matrix }}
 
+      # We use the build for the release, override its name.
+      artifact_archive_name: sycl_windows.tar.gz
+
+  nightly_release:
```
Comment (bader) [inline .github/workflows/sycl_nightly.yml:58]: Let's use "build" instead of "release" everywhere to avoid confusion about quality guarantees.

Comment (aelovikov-intel) [reply .github/workflows/sycl_nightly.yml:58]: I don't think those are user visible. "Build" would also be confusing for the CI developers as technically we are doing "Github release" here and not building anything.

Comment (aelovikov-intel) [reply .github/workflows/sycl_nightly.yml:58]: Can we meet in the middle with "build_prerelease" ?

Comment (bader) [reply .github/workflows/sycl_nightly.yml:58]: Why do you want to call it prerealse? 
Is it just because GitHub doesn't allow anything other than "release" and "pre-release"?
If so, I don't see much sense in calling our nightly builds "pre-release".

Comment (bader) [reply .github/workflows/sycl_nightly.yml:58]: > I don't think those are user visible. 

The next line defines user visible name. 

> "Build" would also be confusing for the CI developers as technically we are doing "Github release" here and not building anything.

"Compiler build upload"?

**Comment by aelovikov-intel** [comment]
Latest testing (before the renaming commit, which I don't plan to test separately):
https://github.com/intel/llvm/actions/runs/5682698878
https://github.com/intel/llvm/actions/runs/5682560302

resulted in the following "release": https://github.com/intel/llvm/releases/tag/nightly-2023-07-27-e329d20

**Comment by aelovikov-intel** [comment]
Latest version tested at https://github.com/intel/llvm/actions/runs/5684780879. Should be ready for the final review round now.

File: devops/containers/ubuntu2204_preinstalled.Dockerfile [hunk 6]
```diff
@@ -5,8 +5,7 @@ FROM $base_image:$base_tag
 
 COPY scripts/drivers_entrypoint.sh /drivers_entrypoint.sh
 RUN mkdir -p /opt/sycl
-ADD llvm_sycl.tar.zst /tmp
-RUN tar -I 'zstd -d' -xf /tmp/llvm_sycl.tar.zst -C /opt/sycl && rm /tmp/llvm_sycl.tar.zst
+ADD sycl_linux.tar.gz /tmp
```
Comment (bader) [inline devops/containers/ubuntu2204_preinstalled.Dockerfile:9]: Why did you remove installation to /opt/sycl?

Comment (aelovikov-intel) [reply devops/containers/ubuntu2204_preinstalled.Dockerfile:9]: See latest commit's message: 

> [Remove unpack - "ADD *.tar.gz /tmp" does it automatically](https://github.com/intel/llvm/pull/10560/commits/2ab670bca7c32bb92fbcd4cf9cadc7b73e71b2af)

It also restores it to what we had before `zstd` compression was introduced in https://github.com/intel/llvm/pull/9675/files#diff-e50cbabbef9e5364e1fbfe55bd3c6385f327ccceeab0e45674913f08afbfb05dR9

Comment (bader) [reply devops/containers/ubuntu2204_preinstalled.Dockerfile:9]: Are you saying that "ADD *.tar.gz /tmp" does "tar -I 'gunzip' -xf /tmp/sycl_linux.tar.gz -C /opt/sycl && rm /tmp/sycl_linux.tar.gz"? I don't think so.

File: devops/containers/ubuntu2204_preinstalled.Dockerfile [hunk 7]
```diff
@@ -5,8 +5,7 @@ FROM $base_image:$base_tag
...
```
Comment (aelovikov-intel) [inline devops/containers/ubuntu2204_preinstalled.Dockerfile:8]:
```suggestion
ADD sycl_linux.tar.gz /opt/sycl/
```

File: devops/containers/ubuntu2204_preinstalled.Dockerfile (hunk 6 above)
Comment (aelovikov-intel) [reply devops/containers/ubuntu2204_preinstalled.Dockerfile]: Almost: https://github.com/intel/llvm/actions/runs/5684595337/job/15407779458

I have to change `/tmp` to `/opt/sycl` but the `tar xf` and `rm` part is automatic. Done in the last commit.

**Review by bader — APPROVED** [review]
Thanks!
LGTM, assuming that unresolved comments will be addressed by follow-up PRs.
